[REFACTOR] Refactor Scheduler following Clean Architecture

anagperal opened the pull request:
### :pushpin: References

-   **Issue:** Closes #8698310pt [Refactor Scheduler following Clean Architecture](https://app.clickup.com/t/8698310pt)

### :memo:  New refactor tasks

- Decouple configRepository from other repositories that only need it to use the dataStoreClient
- Think about how to refactor RepositoryFactory concept
- Do not pass RepositoryFactory and Instance un use cases, only pass the necessary repositories
- In the constructor of data repositories pass instead of Instance we should get D2Api or directly DataStoreClient when different instances are not needed
- CompositionRoot needs to be refactored
- Remove RepositoryConstructor concept from domain repositories where is not neccesary
- Improve and refactor Scheduler tests using ts-mockito and adding integration tests

### :memo: Implementation

- [x] Refactor Scheduler following Clean Architecture: using MVP pattern
- [x] Refactor old code for a better understanding
- [x] Fix feedback component error when init app when not having feedback config in appConfig
- [x] Refactor SchedulerRepository to SchedulerExecutionInfoRepository and use cases that uses this repository
- [x] Decouple configRepository in SchedulerExecutionInfoD2ApiRepository and SyncRuleJobConfigD2ApiRepository
- [x] Fix add lastExecution also to scheduler-executions in datastore
- [x]  First version of Scheduler test --> for the future use ts-mockito
- [x] Add new futures

### :video_camera: Screenshots/Screen capture
![image](https://github.com/user-attachments/assets/4b2689db-c37f-41d0-85c0-1c3a020d787b)

### :fire: Is there anything the reviewer should know to test it?

1. Create or edit a sync rule having the scheduler enabled
2. Run scheduler using the corresponding config: `yarn start-scheduler -c config.json`
3. Check if data is sync correctly

### :bookmark_tabs: Others

-   Any change in the [GUI library](https://github.com/EyeSeeTea/d2-ui-components)? If so, what branch/PR?

-   Any change in the [D2 Api](https://github.com/EyeSeeTea/d2-api)? If so, what branch/PR?

ifoche commented:
Task linked: [CU-8698310pt Refactor Scheduler](https://app.clickup.com/t/8698310pt)

9sneha-n commented on src/data/scheduler/SchedulerExecutionInfoD2ApiRepository.ts:
```diff
@@ -0,0 +1,27 @@
+import { Instance } from "../../domain/instance/entities/Instance";
```
Shall we add a decorator for all refactored files? So that we can keep track of all the files that are pending refactor?

anagperal replied:
I have added comments on the refactored files and TODOs commenting on what they need to be fully refactored. Also in PR I have added some common reactoring tasks

9sneha-n commented on src/data/scheduler/SyncRuleJobConfigD2ApiRepository.ts:
```diff
@@ -0,0 +1,55 @@
+export class SyncRuleJobConfigD2ApiRepository implements SyncRuleJobConfigRepository {
+    private dataStoreClient: StorageDataStoreClient;
+
+    constructor(private instance: Instance) {
```
currently in our projects, we are pass private api: D2Api to all repo constructors and then get the dataStoreClient from it. I am not sure why that is used instead of instance. We can check with Arnau and if that is recommended, shall we start using that pattern in all refactored repos?

anagperal replied:
Code comment and task added for future implementation

9sneha-n commented on src/domain/common/factories/RepositoryFactory.ts:
```diff
@@ -190,9 +191,10 @@ export class RepositoryFactory {
-    public schedulerRepository(instance: Instance) {
-        const config = this.configRepository(instance);
-        return this.get<SchedulerRepositoryConstructor>(Repositories.SchedulerRepository, [config]);
+    public schedulerExecutionInfoRepository(instance: Instance) {
```
do we need to use this concept of repository-factory? I dont see it in our standard apps. We can create a separate PR to handle these "common" refactors, just making a note here for us.

9sneha-n commented on src/domain/scheduler/repositories/SchedulerExecutionInfoRepositoryConstructor.ts:
```diff
@@ -0,0 +1,11 @@
+import { Instance } from "../../instance/entities/Instance";
+import { SchedulerExecutionInfo } from "../entities/SchedulerExecutionInfo";
+
+export interface SchedulerExecutionInfoRepositoryConstructor {
```
Again this comment in for separate PR for "common" refactor - should we used this concept of Constructor ?

9sneha-n reviewed (changes requested):
@anagperal Awesome first PR for refactor. 
Some comments below, we will probably need architect's opinion for some of them, then we can make it a standard and use it in all the next refactor PRs.

anagperal commented on src/domain/scheduler/usecases/GetLastSchedulerExecutionInfoUseCase.ts:
```diff
@@ -0,0 +1,12 @@
+export class GetLastSchedulerExecutionInfoUseCase implements UseCase {
+    constructor(private repositoryFactory: RepositoryFactory, private localInstance: Instance) {}
+
+    public execute(): Promise<SchedulerExecutionInfo> {
```
I have added new Future entity and changed from Promise to Future

anagperal commented on src/domain/scheduler/usecases/GetLastSchedulerExecutionInfoUseCase.ts:
```diff
@@ -0,0 +1,12 @@
+import { SchedulerExecutionInfo } from "../entities/SchedulerExecutionInfo";
+
+export class GetLastSchedulerExecutionInfoUseCase implements UseCase {
+    constructor(private repositoryFactory: RepositoryFactory, private localInstance: Instance) {}
```
Code comment and task added for future implementation

anagperal commented on src/presentation/CompositionRoot.ts:
```diff
@@ -13,7 +13,7 @@ import { GitHubOctokitRepository } from "../data/packages/GitHubOctokitRepositor
 import { FileRulesDefaultRepository } from "../data/rules/FileRulesDefaultRepository";
 import { RulesD2ApiRepository } from "../data/rules/RulesD2ApiRepository";
-import { SchedulerD2ApiRepository } from "../data/scheduler/SchedulerD2ApiRepository";
+import { SchedulerExecutionInfoD2ApiRepository } from "../data/scheduler/SchedulerExecutionInfoD2ApiRepository";
```
Code comment and task added for future implementation

anagperal commented on src/scheduler/SchedulerPresenter.ts:
```diff
@@ -0,0 +1,162 @@
+import { SchedulerExecutionInfo } from "../domain/scheduler/entities/SchedulerExecutionInfo";
+import { Logger } from "./entities/Logger";
+
+export class SchedulerPresenter {
```
Changed!

anagperal commented on src/scheduler/SchedulerPresenter.ts:
```diff
@@ -0,0 +1,162 @@
+
+                // Format date to keep timezone offset
+                const nextDate = moment(job.nextExecution.toISOString()).toISOString(true);
+                this.options.logger.info("scheduler", `Scheduling new sync rule ${name} (${id}) at ${nextDate}`);
```
Done!

anagperal commented on src/scheduler/entities/ScheduledJob.ts:
```diff
@@ -0,0 +1,6 @@
+export type ScheduledJob = {
```
Moved!

anagperal commented on src/scheduler/entities/SchedulerContract.ts:
```diff
@@ -0,0 +1,7 @@
+import { ScheduledJob } from "./ScheduledJob";
+
+export interface SchedulerContract {
```
I would change this together with @xurxodev PR comments because I think this is more related with the architecture

anagperal commented:
> @anagperal Awesome first PR for refactor. Some comments below, we will probably need architect's opinion for some of them, then we can make it a standard and use it in all the next refactor PRs.

Thanks for all the comments @9sneha-n ! I have added comments in code and in the PR for common refactor tasks that we need to do in the future, also I have added them in the whiteboard of MD Sync (CC @adrianq).

xurxodev reviewed (changes requested):
thanks @anagperal 

I leave things to change that we have seen in the live review:
* SyncRuleJobConfigD2ApiRepository and the use case that invoke it doesn't make sense. Better to use ListRulesUseCase passing required filter to avoid duplicate logic to retrive
* Move contracts used by SchedulerCli from domain to near of schedulerCli. This contracts are only used by SchedulerCli and it's contracts of presentation.
* Maybe make sense in the scheduler use cases pass directly the required repository because here the instance is not dynamic
* For tests better to use ts-mockito to create test doubles as we saw in the coding dojo

anagperal commented:
> thanks @anagperal
> 
> I leave things to change that we have seen in the live review:
> 
> * SyncRuleJobConfigD2ApiRepository and the use case that invoke it doesn't make sense. Better to use ListRulesUseCase passing required filter to avoid duplicate logic to retrive
> * Move contracts used by SchedulerCli from domain to near of schedulerCli. This contracts are only used by SchedulerCli and it's contracts of presentation.
> * Maybe make sense in the scheduler use cases pass directly the required repository because here the instance is not dynamic
> * For tests better to use ts-mockito to create test doubles as we saw in the coding dojo

Thanks @xurxodev!! All done except the part about changing the tests to use ts-mockito as you are adding it to another PR. I have noted this task for the future: **Improve and refactor Scheduler tests using ts-mockito and adding integration tests** to also add some integration testing to test syncs between two instances

xurxodev reviewed (approved):
Thanks @anagperal

9sneha-n reviewed (approved):
Thanks @anagperal !
